### ISSUE TITLE
[RFC] Scaffolding anywhere

### ISSUE THREAD

**Issue by sagacity** [backstage/backstage#7211]
**Status:** Open for comments

## Need

Currently scaffolding happens in a separate `create` page. Templates are listed there, users pick a template, they get an input form and scaffolding starts. This covers the use case of 'starting from scratch', for instance to create a new Java project.

I want to propose another use case: 'Adding resources to an existing entity'.

## Proposal

Imagine you already have a working service `service-foo`. We might show a 'messaging' tab showing all pub/sub topics that this service owns. If it does not yet own a pub/sub topic, it would be nice to have a `Add topic` button that allows users to provision a new topic right on that page. What would be the workflow?

* The user clicks 'Add topic'
* An 'add pub/sub topic' template form opens up on the current page
* Since we were already on the `service-foo` page, the user does not need to select `service-foo` via an `EntityPicker` (this needs to be communicated *somehow* to the template)
* The scaffolder runs as normal, doing some provisioning in the cloud to add the pub/sub topic
* If successful, the data on the page may be reloaded so that the list of pub/sub topics is refreshed

In the current scaffolder set up this does not yet seem possible: `TemplatePage` is a separate page that redirects to a `TaskPage` showing scaffolding progress.

My proposal would be to:
* Allow `TemplatePage` to be shown in a modal popup
* Allow additional props to be passed in to `TemplatePage` as defaults for any form elements (in the above example it could pre-populate a hidden `EntityPicker`, for instance)
* Once the scaffolding starts, the task progress is shown in the same popup
* Once the scaffolding is completed, the popup can be closed

## Alternatives

Instead of having the popup an alternative could be to redirect to the `create` page, and via a query string choose the template to use and pre-populate some form elements. Additionally, a `redirect_url` could be used to redirect the user back to where they came from, once scaffolding is complete. This feels simpler from a code point of view, but feels a bit disruptive UX-wise.

Doing the redirect to the `create` page also implies that people could run the "Add pub/sub topic" template directly from the create page. If so, the `EntityPicker` (which chooses which service to add the pub/sub topic to) should be visible. In the proposal I've mentioned this picker should be hidden.

## Risks

* Should there be a distinction between templates runnable from the `create` page versus templates that are runnable as part of any page? This seems messy, so perhaps the rule should be that templates should always be runnable from the `create` page as well.
* In the above proposal and alternative the `EntityPicker` should be hidden in one case (the proposal) and not hidden in another (the alternative). This, too, seems like it could become unclear. It could be covered by some sort of 'hide/read-only if prepopulated' flag.

**Comment by hostsamurai** [comment]
We have a similar idea that is centered around being able to leverage actions
anywhere. Much like this proposal, we were thinking of having either a modal or
a page be a means for kicking off the associated scaffolding tasks. 

In terms of the UI involved, I'm wondering if it would be preferable to
redirect users to the `/create` page rather than have a modal to interact. An
advantage of the modal is that it keeps the user in the current page, so they
don't lose context so to say. A disadvantage is that the modal introduces a new
user interaction pattern that I don't believe exists elsewhere in Backstage. If
it does, please disregard. Otherwise, having this modal would not only increase
the scope of this work but could also require a bit of UX work to get right, as
it could be something that other parts of Backstage could utilize.
